### PR TITLE
Feature/project fugu

### PR DESCRIPTION
**Pull request by lucienimmink**
Add article and talk about Project Fugu
Links to Dave's filesystem article.

### CODE REVIEW

File: data/blog/project-fugu.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+title: 'Project Fugu'
+date: '2023-04-14'
+tags: ['frontend', 'web-capabilities', 'project-fugu']
+images: ['/articles/project-fugu/header.webp']
+summary: Cross-platform development is hard. Each platform has its specific implementation of an API and you end up with separate apps for each platform. Multiple frameworks try to fix this issue by creating an abstract between the platform and the application. Browser vendors are doing the same and it's called Project Fugu.
+authors: ['lucien-immink']
+theme: 'rouge'
+---
+
+Cross-platform development is hard. Features and capabilities of an application require you to think about how it needs to be implemented in different platforms like Android, iOS, Web, Windows, MacOS and Linux. If you, for example, need to access the address details of a contact that is stored on a device you need to implement a contact picker for all the different platforms you want to support or you need to come up with a completely custom implementation. Custom implementations are probably not what you want to do. It can become quite messy quite fast.
```
Comment (mrtnvh) [inline data/blog/project-fugu.md]: Cross-platform _software_ development?

Comment (mrtnvh) [inline data/blog/project-fugu.md]: to be implemented _on_ different platforms?

Comment (mrtnvh) [inline data/blog/project-fugu.md]: A little bit more information about what a contact picker is?

...to implement a piece of software, that allows you to access and choose a contact, let's call it a contact picker, for all the different platforms...

File: data/blog/project-fugu.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+So what can you do?
+
+What about an abstraction layer between the OS native API and a common top layer, which is preferably written in a well-known set of languages that can handle UIs with a breeze, like HTML, CSS and JavaScript? Here we are describing a framework like Apache Cordova, previously known as PhoneGap. These _hybrid_ applications combine web languages with a wrapper for the OS native APIs. Hybrid cause they use web technology packaged as apps for distribution and have access to native device APIs.
```
Comment (mrtnvh) [inline data/blog/project-fugu.md]:
```suggestion
What about an abstraction layer between the OS native API and a common top layer, which is preferably written in a well-known set of languages that can handle UIs with a breeze, like HTML, CSS and JavaScript? Here we are describing a framework like Apache Cordova, Ionic's Capacitor, Electron or Tauri. These _hybrid_ applications combine web languages with a wrapper for the OS native APIs. Hybrid cause they use web technology packaged as apps for distribution and have access to native device APIs.
```

File: data/blog/project-fugu.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+While technology like Apache Cordova enables developers to build one UI for multiple platforms they still have to build the application for all the platforms they want to support and sometimes the support for a platform or specific feature is a bit flaky.
```
Comment (mrtnvh) [inline data/blog/project-fugu.md]:
```suggestion
While these hybrid frameworks enable developers to build one UI for multiple platforms they still have to build the application for all the platforms they want to support. The support for a platform or platform specific feature can often be flaky.
```

_Often_ because I am speaking from experience 😄

File: data/blog/project-fugu.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+Project Fugu 🐡 is an effort to close gaps in the web's capabilities enabling new classes of applications to run on the web. APIs that Project Fugu is delivering enable new experiences on the web while preserving the web's core benefits of security, low friction, and cross-platform delivery. All Project Fugu API proposals are made in the open and on the standards track.
+
+![Schema of the elements of a URL](/articles/project-fugu/api-call-abstraction.webp)
+
+Fugu is organized as a Chromium project, open to all Chromium contributors and organizations. Today, that includes Microsoft, Intel, Samsung, Electron and Google (among others). Fugu Leads triage incoming requests from partners, determine demand, prioritize them, find champions, track development, and help organize the release and documentation for the capability.
+
+![Schema of the elements of a URL](/articles/project-fugu/adding-apis.webp)
+
+Let's take a look at some of the features that have already been released:
```
Comment (mrtnvh) [inline data/blog/project-fugu.md:31]: IMHO, I think this could use an additional h2, right above this sentence.

## Fugu features already available

File: data/blog/project-fugu.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+Support for webOTP has been added since Chrome 93, Edge 93. Firefox and Safari have no support, meaning that webOTP is not readily available yet.
+
+### File System Access
+
+[Dave Bitter](https://techhub.iodigital.com/authors/dave-bitter) has written an excellent article about [File system access](https://techhub.iodigital.com/articles/the-file-system-access-api)
+
+## Features in the pipeline
+
+Some of the features that are being developed or considered at the moment
```
Comment (mrtnvh) [inline data/blog/project-fugu.md]:
```suggestion
Some of the features that are being developed or considered at the moment:
```

File: data/blog/project-fugu.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+- Changing system settings API
+- Geofencing
+- Splash screen
+- Remote desktop control
+- Call dialer/answering/control
+
+## ...but what about the other one?
+
+Project Fugu can offer a great experience, but only if the browser supports these APIs. Safari (and thus iOS and iPad OS) are not actively part of project fugu and have been slow with implementing APIs that extend the functionality of the web, like service worker and push notifications. Perhaps that will change, but it might not. Who knows if Apple will ever allow (is forced to) other browser engines on iOS and iPadOS. Having APIs fully specced and tested has a positive effect on browsers without an implementation since they can follow what has been laid out before them.
```
Comment (mrtnvh) [inline data/blog/project-fugu.md]:
```suggestion
Project Fugu can offer a great experience, but only if the browser supports these APIs. Safari (and thus iOS and iPad OS) are not actively part of Project Fugu and have been slow with implementing APIs that extend the functionality of the web. Eg. support for service workers and push notifications. 

Perhaps that will change, but it might not. Who knows if Apple will ever allow (is forced to) other browser engines on iOS and iPadOS. 

Having APIs fully specified and tested has a positive effect on browsers without an implementation since they can follow what has been laid out before them.
```

Comment (mrtnvh) [inline data/blog/project-fugu.md]: IMHO, I think the phrase `slow with implementing APIs that extend the functionality of the web`

Although SW and Push support have been a long time coming, there are clear indications that the Safari team is moving the web platform forward. Look at https://wpt.fyi/interop-2023

Maybe rephrase this sentence with more information? I think you are aware what forces are currently in play, eg. EU regulations.

For example:

> If Safari and Apple will ever actively contribute to Project Fugu, still remains an open question. On the one hand, the fact that Apple does not allow other browsers engines from their mobile platforms, is one of the main reasons why web apps are still considered less than a native app. On the other, EU regulators have noticed this as well and are actively in debate if this 'browser ban' is considered to be legal. For more information concerning this specific topic, check out https://open-web-advocacy.org/

File: data/blog/project-fugu.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+So what can you do?
+
+Progressive Web Apps (PWAs) bring offline support and _app-like_ experiences to the web. Extending the PWA principle with APIs to include more native APIs while keeping the principles of the web (trust, privacy and security) is the core of Project Fugu 🐡.
+
+Some of these APIs are readily available and should be considered when that feature is needed, like for example the async clipboard API. Other APIs though are only available in certain browsers, like the webOTP API so handle them with care and consult the online references.
+
+Who knows if Apple will ever allow (is forced to) other browser engines on iOS and iPadOS. Having APIs fully specced and tested has a positive effect on browsers without an implementation since they can follow what has been laid out before them.
```
Comment (mrtnvh) [inline data/blog/project-fugu.md]: I'd like to refer to my previous comment about Safari and Apple.

File: data/jobs.json [hunk 8]
```diff
@@ -1 +1,1661 @@
...le":"Drupal Developer","location":"Utrecht, Netherlands","country":"Netherlands","city":"Utrecht","postal_code":"","country_code":"NL","department":"Technology","employment_type":"fulltime","category":"internet","experience":"mid_level","education":"bachelor_degree","tags":{"tag":["PHP","React","Drupal 8","REST","GraphQL"]},"min_hours":32,"max_hours":40,"careers_url":"https://iodigital.com/nl/carriere/vacatures/drupal-developer","created_at":"2001-02-01T00:00:00.000Z","updated_at":"2001-02-01T00:00:00.000Z","published_at":"2001-02-01T00:00:00.000Z","closed":""},{"id":"812022","title":"Mobile Developer React Native","location":"Den Bosch, Netherlands","country":"Netherlands","city":"Den Bosch","postal_code":"5222 AE","country_code":"NL","department":"Technology","employment_type":"fulltime","category":"internet","experience":"mid_level","education":"bachelor_degree","tags":{"tag":["React Native","Javascript / Typescript","React"]},"min_hours":32,"max_hours":40,"careers_url":"https://iodigital.com/nl/carriere/vacatures/mobile-developer-react-native","created_at":"2001-02-01T00:00:00.000Z","updated_at":"2001-02-01T00:00:00.000Z","published_at":"2001-02-01T00:00:00.000Z","closed":""}]}
```
Comment (mrtnvh) [inline data/jobs.json:1]: Why is this file changed? Should this be the case?

**Review by mrtnvh — CHANGES_REQUESTED** [review]
Hi Lucien,

I've added some questions and suggestions that came to mind.

Why have you changed data/jobs.json and the lockfile? I don't think they should be part of this feature/

File: data/jobs.json (hunk 8 above)
Comment (lucienimmink) [reply data/jobs.json:1]: I merged everything, I'll remove this file since I am not trying to change our job offerings :D

Comment (sanderdejong88) [reply data/jobs.json:1]: You can keep it, that's fine. It automatically updates twice a week.

File: data/blog/project-fugu.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+So what can you do?
+
+What about an abstraction layer between the OS native API and a common top layer, which is preferably written in a well-known set of languages that can handle UIs with a breeze, like HTML, CSS and JavaScript? Here we are describing a framework like Apache Cordova, Ionic's Capacitor, Electron or Tauri. These _hybrid_ applications combine web languages with a wrapper for the OS native APIs. Hybrid cause they use web technology packaged as apps for distribution and have access to native device APIs.
+
+While these hybrid frameworks enable developers to build one UI for multiple platforms they still have to build the application for all the platforms they want to support. The support for a platform or platform specific feature can often be flaky.
```
Comment (sanderdejong88) [inline data/blog/project-fugu.md:17]: Isn't it `platform-specific` instead of `platform specific`?

**Comment by mrtnvh** [comment]
LGTM

File: data/blog/project-fugu.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+---
+
+Cross-platform software development is hard. Features and capabilities of an application require you to think about how it needs to be implemented on different platforms like Android, iOS, Web, Windows, MacOS and Linux. If you, for example, need to access the address details of a contact that is stored on a device you need to implement a piece of software, that allows you to access and choose a contact, let's call it a contact picker, for all the different platforms you want to support or you need to come up with a completely custom implementation. Custom implementations are probably not what you want to do. It can become quite messy quite fast.
+
+So what can you do?
+
+What about an abstraction layer between the OS native API and a common top layer, which is preferably written in a well-known set of languages that can handle UIs with a breeze, like HTML, CSS and JavaScript? Here we are describing a framework like Apache Cordova, Ionic's Capacitor, Electron or Tauri. These _hybrid_ applications combine web languages with a wrapper for the OS native APIs. Hybrid cause they use web technology packaged as apps for distribution and have access to native device APIs.
```
Comment (sanderdejong88) [inline data/blog/project-fugu.md:15]: Maybe link the frameworks you're mentioning?
